crates/pathfinder/src/core.rs

```diff
@@ -3,53 +3,86 @@
 //!
 //! This includes many trivial wrappers around [StarkHash] which help by providing additional type safety.
 use pedersen::StarkHash;
+use serde::{Deserialize, Serialize};
 use web3::types::H256;
 
 /// The address of a StarkNet contract.
-#[derive(Debug, Copy, Clone, PartialEq)]
+#[derive(Debug, Default, Copy, Clone, PartialEq, Deserialize, Serialize)]
 pub struct ContractAddress(pub StarkHash);
```

Comment:
Why would this get a default?

Reply:
Same for all the rest..

Reply:
Similar for serde stuff. There are more than 1 way to serde these types so we should be very careful adding something this broad.

Reply:
Defaults were added as a consequence of the latter issue described below, ie. how to represent missing fields for the pending blocks in the rpc api, and other fields that can simply be missing in the sequencer replies (not to confuse with invalid values which should be treated as an error, as Joonas pointed out earlier).

Reply:
As for serde - I think the most common way to serialize is into hex string, which is the only way specified in the RPC spec.
As for deserialization - this is also the most common way to deserialize, looking at data coming from the sequencer.
As for deserialization - other forms of input to deserialize into stark hash based types can be done through specialized functions and proper serde annotations.
This way you get:
- serialization boilerplate for newtypes nonexistent
- deserialization boilerplate for newtypes nonexistent
- deserialization for rare long uints and other peculiar cases managed by serde annotations.

The alternative (ie. dropping Serialize and Deserialize from StarkHash and in consequence dropping their derives in newtypes) means that we either need to add a lot of redundant boilerplate for Serialize/Deserialize impls per newtype or add serde (deserialize_with/serialize_with) annotations for each and every field that is of some newtype and happens to be deserialized from sequencer replies or serialized into rpc replies.

Reply:
I would rather stick to the current, specific solution wrt serde instead of being too orthodox wrt various theoretical formats to serialize to/deserialize from (which would be implementing mostly non existent use cases right now).

Reply:
> At the moment I assumed zero hash for such values 

Then set it zero manually.

Reply:
Is there not a serde tag thingy to use a specific function for the serialization? That's why I thought we wanted the `to_hex_str` for all these new-types.

Reply:
> Then set it zero manually.
I will check with Starkware concerning missing values and how to represent them in rpc output :+1:

Reply:
> Is there not a serde tag thingy to use a specific function for the serialization? That's why I thought we wanted the to_hex_str for all these new-types.
This actually has some caveats:
- if `to_hex_str` had a typical signature of `fn to_hex_str(&self) -> String` then we would actually slow down the serialization 5-10 times just because of this intermediate step of having a heap allocated representation which is not really necessary for serialization as such
- so actually I wouldn't use such a function for serialization and instead use some standalone one that works on a mut slice and mimics the current `StarkHash::serialize` impl
- just like I mentioned before - in order to keep boilerplate to minimum and `StarkHash` not implementing `Serialize` directly I would have to use either `serde(serialize_with = "...")` or `serde_as(`as = "..."`)` depending on which solution keeps boilerplate code for each and every newtype to a minimum, this ofc also means annotating each and every field, which is now avoided